### PR TITLE
sheng/propagate decision reason

### PR DESCRIPTION
**Pull request by sheng-stripe**
The problem here is that the classification for an action's decision isn't propagated properly, and so we're getting some incorrect decision_reasons logged out to splunk. Tests all pass, but I'm not sure if I've covered all the corner cases.

r? @ajj-stripe @tremblay-stripe

### CODE REVIEW

File: pkg/smokescreen/egressacl.go [hunk 1]
```diff
@@ -1,6 +1,6 @@
 package smokescreen
 
 type EgressAcl interface {
-	Decide(fromService string, toHost string) (EgressAclDecision, bool, error)
+	Decide(fromService string, toHost string) (EgressAclDecision, string, bool, error)
```
Comment (ajj-stripe) [inline pkg/smokescreen/egressacl.go]: I think it's worth documenting here and in the implementation function header what each of the return values represent.

File: pkg/smokescreen/smokescreen.go [hunk 2]
```diff
@@ -123,21 +123,21 @@ func classifyAddr(config *Config, addr *net.TCPAddr) ipType {
...
 
 	classification := classifyAddr(config, resolved)
 	config.StatsdClient.Incr(classification.statsdString(), []string{}, 1)
 
 	if classification.IsAllowed() {
-		return resolved, nil
+		return resolved, classification.String(), nil
 	}
-	return nil, denyError{fmt.Errorf("The destination address (%s) was denied by rule '%s'", resolved.IP, classification)}
+	return nil, "", denyError{fmt.Errorf("The destination address (%s) was denied by rule '%s'", resolved.IP, classification)}
```
Comment (ajj-stripe) [inline pkg/smokescreen/smokescreen.go]: If `!classification.IsAllowed()`, the egress will be blocked. In that case, we want to populate the decision reason with something similar to the error message supplied here, right? Because we want to be able to say something like "Denied because resolved IP matched rule X"

File: pkg/smokescreen/smokescreen.go [hunk 3]
```diff
@@ -123,21 +123,21 @@ func classifyAddr(config *Config, addr *net.TCPAddr) ipType {
...
-		return nil, err
+		return nil, "", err
 	}
 
 	classification := classifyAddr(config, resolved)
 	config.StatsdClient.Incr(classification.statsdString(), []string{}, 1)
 
 	if classification.IsAllowed() {
-		return resolved, nil
+		return resolved, classification.String(), nil
```
Comment (sheng-stripe) [inline pkg/smokescreen/smokescreen.go]: String is overloaded for classification, which is an enum. It provides useful information informing why an IP was allowed or denied which we should propagate up.

File: pkg/smokescreen/acl_loader_v1.go [hunk 4]
```diff
@@ -18,41 +18,41 @@ type EgressAclRule struct {
...
 
-func (ew *EgressAclConfig) Decide(fromService string, toHost string) (EgressAclDecision, bool, error) {
+func (ew *EgressAclConfig) Decide(fromService string, toHost string) (EgressAclDecision, string, bool, error) {
 	var action EgressAclDecision
 
 	rule := ew.ruleForService(fromService)
 	if rule == nil {
 		action = EgressAclDecisionDeny
-		return action, false, nil
+		return action, "", false, nil
```
Comment (ajj-stripe) [inline pkg/smokescreen/acl_loader_v1.go]: Worth supplying a reason here, right? "No rule matched" or similar?

File: pkg/smokescreen/smokescreen.go [hunk 5]
```diff
@@ -577,20 +577,20 @@ func checkIfRequestShouldBeProxied(config *Config, req *http.Request, outboundHo
 
 	switch action {
 	case EgressAclDecisionDeny:
-		decision.reason = "Role is not allowed to access this host"
```
Comment (sheng-stripe) [inline pkg/smokescreen/smokescreen.go]: Previously, this would in certain cases this decision reason would be incorrect, which it should be something like ""Deny: User Configured".

File: pkg/smokescreen/acl_loader_v1.go [hunk 6]
```diff
@@ -18,41 +18,41 @@ type EgressAclRule struct {
...
 	}
 
 	defaultRuleUsed := rule == ew.Default
 
 	// if the host matches any of the rule's allowed domains, allow
 	for _, domainGlob := range rule.DomainGlob {
 		if hostMatchesGlob(toHost, domainGlob) {
-			return EgressAclDecisionAllow, defaultRuleUsed, nil
+			return EgressAclDecisionAllow, "if the host matches any of the rule's allowed domains, allow", defaultRuleUsed, nil
```
Comment (ajj-stripe) [inline pkg/smokescreen/acl_loader_v1.go]: I think the phrasing for these should resemble more in the active past tense, like "Destination matched a rule in the role's list of allowed_domains"

(same for the next few reason strings)

File: pkg/smokescreen/smokescreen.go [hunk 7]
```diff
@@ -356,7 +356,7 @@ func handleConnect(config *Config, ctx *goproxy.ProxyCtx) (*net.TCPAddr, error)
 		return nil, denyError{errors.New(decision.reason)}
 	}
 
-	resolved, err = safeResolve(config, "tcp", ctx.Req.Host)
+	resolved, decision.reason, err = safeResolve(config, "tcp", ctx.Req.Host)
```
Comment (sheng-stripe) [inline pkg/smokescreen/smokescreen.go]: My assumption is that should mutate the decision object in go, but uhh this is my first time writing golang. D:

File: pkg/smokescreen/acl_loader_v1.go [hunk 8]
```diff
@@ -62,16 +62,16 @@ func (ew *EgressAclConfig) Decide(fromService string, toHost string) (EgressAclD
 	case ConfigEnforcementPolicyEnforce:
 		action = EgressAclDecisionDeny
 	case ConfigEnforcementPolicyOpen:
-		return EgressAclDecisionAllow, defaultRuleUsed, nil
+		return EgressAclDecisionAllow, "", defaultRuleUsed, nil
```
Comment (ajj-stripe) [inline pkg/smokescreen/acl_loader_v1.go]: We can supply a reason here, like "Role has open enforcement policy"

File: pkg/smokescreen/acl_loader_v1.go [hunk 9]
```diff
@@ -62,16 +62,16 @@ func (ew *EgressAclConfig) Decide(fromService string, toHost string) (EgressAclD
...
 	default:
-		return 0, defaultRuleUsed, fmt.Errorf("unexpected policy value for (%s -> %s): %d", fromService, toHost, rule.Policy)
+		return 0, "", defaultRuleUsed, fmt.Errorf("unexpected policy value for (%s -> %s): %d", fromService, toHost, rule.Policy)
 	}
 
 	// use the decision from rule.Policy
-	return action, defaultRuleUsed, nil
+	return action, "", defaultRuleUsed, nil
```
Comment (ajj-stripe) [inline pkg/smokescreen/acl_loader_v1.go]: On line 62 we can also assign a reason, and return it alongside the action that was selected here. That way when we hit this case, we can log which policy was selected, and that the domain fell through to the policy action.

**Comment by tremblay-stripe** [comment]
Hi, I'm pretty busy. I defer to @ajj-stripe's judgement. You could also ask @mdoan-stripe. I'm still happy to help in second line!

File: pkg/smokescreen/acl_loader_v1_test.go [hunk 10]
```diff
@@ -190,7 +190,7 @@ func TestLoadFromYaml(t *testing.T) {
 		a.Equal(0, len(acl.GlobalAllowList))
 	}
 
-    // Load a sane config with global lists
+	// Load a sane config with global lists
```
Comment (ajj-stripe) [inline pkg/smokescreen/acl_loader_v1_test.go]: We should add tests in this file to ensure the correct reason string is supplied for each case.
- Blocked because it matches an ip rule
- Blocked due to enforce mode
- Blocked due to global deny list
- Allowed due to `allowed_domain` presence
- Allowed due to `open` or `report` mode
- Allowed due to `global_allow_list`

File: pkg/smokescreen/smokescreen.go (hunk 7 above)
Comment (ajj-stripe) [reply pkg/smokescreen/smokescreen.go]: Yeah, this should do what you expect under my understanding of golang.

File: pkg/smokescreen/acl_loader_v1_test.go (hunk 10 above)
Comment (sheng-stripe) [reply pkg/smokescreen/acl_loader_v1_test.go]: Thanks will do. Wanted to throw this up to see if I had the right idea.

File: pkg/smokescreen/acl_loader_v1.go (hunk 6 above)
Comment (sheng-stripe) [reply pkg/smokescreen/acl_loader_v1.go]: sgtm, will do.

File: pkg/smokescreen/egressacl.go (hunk 1 above)
Comment (sheng-stripe) [reply pkg/smokescreen/egressacl.go]: Sure.

File: pkg/smokescreen/smokescreen.go (hunk 2 above)
Comment (sheng-stripe) [reply pkg/smokescreen/smokescreen.go]: Ah, it was my assumption that we'll log out errors when they occur. Is that wrong?

Seems like currently blank strings get set for reason when an error is detected.

File: pkg/smokescreen/smokescreen.go (hunk 7 above)
Comment (sheng-stripe) [reply pkg/smokescreen/smokescreen.go]: Cool, thanks.

File: pkg/smokescreen/smokescreen.go (hunk 2 above)
Comment (ajj-stripe) [reply pkg/smokescreen/smokescreen.go]: We do log out errors, but it's still worth making the `decision_reason` accurate.
I don't think `decision_reason` is cleared when an error is present, but maybe it is. In that case, we should maybe supply `error encountered, see error field` or something as the `decision_reason`.

I think it's probably a blank string because in the cases where we supply errors, we're not in the correct context to modify/provide a `decision_reason`.

File: pkg/smokescreen/acl_loader_v1_test.go (hunk 10 above)
Comment (sheng-stripe) [reply pkg/smokescreen/acl_loader_v1_test.go]: I'll add some more tests for ip rules.

Comment (sheng-stripe) [reply pkg/smokescreen/acl_loader_v1_test.go]: Eh, actually. Not sure if there's an easy way to mock this.

**Comment by sheng-stripe** [comment]
@ajj-stripe re-r?

**Comment by ajj-stripe** [comment]
lgtm, would like to get a more smokescreen-knowledgeable review from rlk

r? @rlk-stripe

File: pkg/smokescreen/smokescreen.go [hunk 11]
```diff
@@ -148,7 +148,7 @@ func dial(config *Config, network, addr string, userdata interface{}) (net.Conn,
 		outboundHost = v.decision.outboundHost
 	}
 
-	resolved, err := safeResolve(config, network, addr)
+	resolved, _, err := safeResolve(config, network, addr)
```
Comment (rlk-stripe) [inline pkg/smokescreen/smokescreen.go]: I think we want to update the decision reason here, similar to line 359.  (This code path is hit in the case where we're proxying plain HTTP rather than HTTPS or other traffic via CONNECT.)

**Comment by sheng-stripe** [comment]
@rlk-stripe re-r? Thanks!
